kernel/src/kernel.rs

```diff
@@ -873,188 +865,206 @@ impl Kernel {
-                subdriver_number,
-                upcall_ptr,
-                appdata,
-            } => {
-                // A upcall is identified as a tuple of the driver number and
-                // the subdriver number.
-                let upcall_id = UpcallId {
-                    driver_num: driver_number,
-                    subscribe_num: subdriver_number,
-                };
-
-                // First check if `upcall_ptr` is null. A null `upcall_ptr` will
-                // result in `None` here and represents the special
-                // "unsubscribe" operation.
-                let ptr = NonNull::new(upcall_ptr);
-
-                // For convenience create an `Upcall` type now. This is just a
-                // data structure and doesn't do any checking or conversion.
-                let upcall = Upcall::new(process.processid(), upcall_id, appdata, ptr);
-
-                // If `ptr` is not null, we must first verify that the upcall
-                // function pointer is within process accessible memory. Per
-                // TRD104:
-                //
-                // > If the passed upcall is not valid (is outside process
-                // > executable memory...), the kernel...MUST immediately return
-                // > a failure with a error code of `INVALID`.
-                let rval1 = ptr.map_or(None, |upcall_ptr_nonnull| {
-                    if !process.is_valid_upcall_function_pointer(upcall_ptr_nonnull) {
-                        Some(ErrorCode::INVAL)
-                    } else {
-                        None
-                    }
-                });
-
-                // If the upcall is either null or valid, then we continue
-                // handling the upcall.
-                let rval = match rval1 {
-                    Some(err) => upcall.into_subscribe_failure(err),
-                    None => {
-                        // At this point we must save the new upcall and return
-                        // the old. The upcalls are stored by the core kernel in
-                        // the grant region so we can guarantee a correct upcall
-                        // swap. However, we do need help with initially
-                        // allocating the grant if this driver has never been
-                        // used before.
+            Syscall::Exit {
+                which,
+                completion_code,
+            } => match which {
+                // The process called the `exit-terminate` system call.
+                0 => process.terminate(Some(completion_code as u32)),
+                // The process called the `exit-restart` system call.
+                1 => process.try_restart(Some(completion_code as u32)),
+                // The process called an invalid variant of the Exit
+                // system call class.
+                _ => process.set_syscall_return_value(SyscallReturn::Failure(ErrorCode::NOSUPPORT)),
+            },
```

Comment:
Why did this move here?

Reply:
I was hoping that Rust would be smart enough to determine that https://github.com/tock/tock/blob/3ecb382cec9684052fc373087c2bf2f070cfd843/kernel/src/kernel.rs#L1365 is not possible and not necessary. I will move it back.